### PR TITLE
Feature/scheduler changes

### LINKED ISSUE #256: Consider changes to Scheduler's updating of future Hours?
Sheryl Stauss (Intercongregation Communities Association Inc.) hub post here: https://powerofus.force.com/0D580000038Pp9F

When the scheduler runs and looks at future Shifts and Hours, it reuses them if they match the correct schedule, but it updates some of the fields to match the potentially modified JRS or VRS.  Specifically, I think Sheryl's hours are being updated based on this rule:

"Future Hours that match the VRS are reused and have the following fields updated to match the VRS: Hours Worked, Number of Volunteers, Status, Planned Start Date & Time, and Comments."

I think the scenario points to a valid case where the VRS should not update the Status if the Status is already set to Canceled.  But what about other fields on the VRS? when should they get updated or not touched?  Can we easily tell the difference between the user modifying the VRS, in which case you'd expect it to update most/all fields on the Hours, vs. the scheduler reverifying future hours?

@mpusto we need to go through this together to come up with any refinement we want to make to the rules as described here (internal doc): https://docs.google.com/document/d/1TwiYpVufmGie9x7nO7bNtWLw_ENP-kKU-yPP8xU3Z44/edit#

here's a copy for public viewing: 
[RecurrenceScheduleProcessing.pdf](https://github.com/SalesforceFoundation/Volunteers-for-Salesforce/files/743676/RecurrenceScheduleProcessing.pdf)

### PR DESCRIPTION
**Pull request by davidhabib**
# Critical Changes

# Changes
- Processing Job Recurrence Schedules has been modified to no longer delete past or future Shifts that have Hours of any status associated with them. 
- The new report "Shifts Detached from their JRS" lists these Shifts that no longer match their JRS's schedule, but were not deleted.
- Processing Volunteer Recurrence Schedules has been modified to no longer any delete past Hours, or future Hours with status of Completed or Canceled.
- The new report "Hours Detached from the VRS" lists these Hours that no longer match their VRS's schedule, but were not deleted.

# Issues Closed
Fixes #256

### CODE REVIEW

**Comment by davidhabib** [comment]
@jessingrass @judisohn I would really appreciate one of you grabbing this pull request branch and giving it a thorough test to make sure the behavior of modifying or deleting JRS's and VRS's matches the behavior we agreed to described here: https://docs.google.com/document/d/1TwiYpVufmGie9x7nO7bNtWLw_ENP-kKU-yPP8xU3Z44/edit#

**Comment by davidhabib** [comment]
@judisohn Could you also review the two new reports I provided, to make sure they have the output format you think would work best.  I specifically did them with no grouping to make it easy to export to excel to get Shift Id's or Hour Id's, so the admin can use them in a bulk update operation.

**Comment by judisohn** [comment]
@davidhabib I went through everything in doc, bullet by bullet and the 2 reports and everything appears to be working exactly it should. Looks great to me. Thanks!

**Comment by davidhabib** [comment]
@njjc Hi Nic, can you please take a look at this soon, so I have a chance to respond to any changes?  thanks.

File: src/classes/VOL_JRS.cls [hunk 1]
```diff
@@ -35,26 +35,64 @@
...
+    private static void calcDeleteOrDetachJRSShifts(
+        list<Volunteer_Shift__c> listShift, 
+        list<Volunteer_Shift__c> listShiftDelete, 
+        list<Volunteer_Shift__c> listShiftUpdate) {
+
+        for (Volunteer_Shift__c shift : listShift) {
+            Integer count = 0;
+            // note that to avoid query limits, we avoid looking at Volunteer_Hours__r.size(), and 
+            // must leverage soql for loops that internally use queryMore. 
+            for (Volunteer_Hours__c hr : shift.Volunteer_Hours__r) {
```
Comment (njjc) [inline src/classes/VOL_JRS.cls]: Do you know of any documentation on this behavior? Is it just for subqueries? My understanding is that each row of the subquery counts against the query row governor limit, not just rows that are accessed. @davidhabib

Comment (njjc) [inline src/classes/VOL_JRS.cls]: I was hoping there would be some standard rollups of hours to shifts, so we could just use that field and avoid the subqueries altogether, but alas, there aren't any.

Comment (davidhabib) [inline src/classes/VOL_JRS.cls]: i stumbled across it in the Apex documentation under processing sub selects.

**Pull request merged**